### ISSUE TITLE
FSDP vs Deepspeed

### ISSUE THREAD

**Issue by ajseo95** [huggingface/trl#1224]
Hi, I have a question about the performance difference between FSDP and Deepspeed zero3.

When I trained with DPO on the SFT model, these two showed similar performance on some NLP datasets(truthfulqa, mmlu, etc.)

However, when training SFT with ultrachat-200k dataset, using FSDP and Deepspeed shows a big performance gap:
<meta charset="utf-8"><div dir="ltr" style="margin-left:0pt;" align="left">
 | LLaMA2 | LLaMA2 SFT deepspeed | LLaMA2 SFT fsdp
-- | -- | -- | --
hellaswag | 76 | 76.72 | 55.95
truthfulqa | 38.96 | 41.61 | 45.24
mmlu | 41.77 | 41.40 | 28.29
</div></b>
Also, training loss converged when using Deepspeed whereas it increased and sometimes vibrated with FSDP.

It seems like underfitting when training with deepspeed.
Is this right? Why does this happen? Is there anyone who has similar issues?

Thank you for your help in advance!

**Comment by younesbelkada** [comment]
Hi @ajseo95 
the slight differences between normal SFT and DS can makes sense to me, however the gap between FSDP and DS is indeed strange. Can you elaborate more on the FSDP config you are using? Also are you training adapters or full fine-tuning?

**Comment by younesbelkada** [comment]
Hi @ajseo95 
This might be due to the way you load your model - do you load your model in full precision ? (`torch_dtype=torch.float32`), I suggest you to have a look at this FSDP training script from @pacman100 : https://github.com/pacman100/DHS-LLM-Workshop/blob/main/chat_assistant/training/train.py who recently introduced a fix on the SFTTrainer: https://github.com/huggingface/trl/pull/1110 to make sure non-trainable params are upcasted in fp32. Can you make sure your model is loaded entirely in fp32?

**Comment by ajseo95** [comment]
I set torch dtpye as bfloat16. I'll try with fp32! Thank you :)
